clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashCollection.java

```diff
@@ -144,18 +156,119 @@ public E next() {
             }
             cur = next;
             next = indexToElement(head, elements, cur.next());
+            ++cursor;
             @SuppressWarnings("unchecked")
             E returnValue = (E) cur;
+            lastReturned = returnValue;
             return returnValue;
         }
 
         @Override
-        public void remove() {
+        public E previous() {
             if (cur == head) {
+                throw new NoSuchElementException();
+            }
+            next = cur;
+            cur = indexToElement(head, elements, cur.prev());
+            --cursor;
+            @SuppressWarnings("unchecked")
+            E returnValue = (E) next;
+            lastReturned = returnValue;
+            return returnValue;
+        }
+
+        @Override
+        public int nextIndex() {
+            return cursor;
+        }
+
+        @Override
+        public int previousIndex() {
+            return cursor - 1;
+        }
+
+        @Override
+        public void remove() {
+            if (lastReturned == null) {
                 throw new IllegalStateException();
             }
-            ImplicitLinkedHashSet.this.remove(cur);
-            cur = head;
+
+            int prev = lastReturned.prev();
+            ImplicitLinkedHashCollection.this.remove(lastReturned);
+            if (lastReturned == cur) {
+                cursor--;
+                cur = indexToElement(head, elements, prev);
+            } else {
+                next = indexToElement(head, elements, prev);
+            }
+
+            if (prev == HEAD_INDEX) {
+                lastReturned = null;
+            }
+        }
+
+        @Override
+        public void set(E e) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void add(E e) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private class ImplicitLinkedHashCollectionListView extends AbstractSequentialList<E> {
+
+        @Override
+        public ListIterator<E> listIterator(int index) {
+            if (index < 0 || index > size) {
```

Comment:
should be index >= size, right?

Reply:
No, this is right. A ListIterator has `size + 1` valid cursor positions. `listIterator(size)` should return an iterator at the end of the list that can move backwards but not forwards.

Reply:
Good call-- thanks for the correction.